onnxruntime/core/framework/tensor_type_and_shape.h

```diff
@@ -2,25 +2,50 @@
 // Licensed under the MIT License.
 #pragma once
 
+#include <memory>
 #include <string>
 #include <vector>
 
 #include "core/framework/tensor_shape.h"
 #include "core/session/onnxruntime_c_api.h"
 
+namespace ONNX_NAMESPACE {
+class TypeProto;
+}
+
+namespace onnxruntime {
+class DataTypeImpl;
+}
+
 struct OrtTensorTypeAndShapeInfo {
  public:
+  using Ptr = std::unique_ptr<OrtTensorTypeAndShapeInfo>;
+
   ONNXTensorElementDataType type = ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT;
   onnxruntime::TensorShape shape;
   // dim_param values. empty string if dim_value or no dim_param was specified.
   // one entry per dimension in shape. only guaranteed to be populated for graph inputs and outputs
   std::vector<std::string> dim_params;
 
   OrtTensorTypeAndShapeInfo() = default;
-  OrtTensorTypeAndShapeInfo(const OrtTensorTypeAndShapeInfo& other) = delete;
-  OrtTensorTypeAndShapeInfo& operator=(const OrtTensorTypeAndShapeInfo& other) = delete;
+  ~OrtTensorTypeAndShapeInfo();
+
+  Ptr Clone() const {
+    return std::make_unique<OrtTensorTypeAndShapeInfo>(*this);
+  }
+
+  OrtTensorTypeAndShapeInfo(const OrtTensorTypeAndShapeInfo& other) = default;
```

Comment:
Did you intend to mark them as deleted?

Reply:
No, the code is correct. Clone() in this case makes use of the copy __ctor. It is a code pattern to use Clone in TypeInfo impl.

Reply:
Not following. What's the reason for undeleting the copy ctor and copy assignment operator? If they need to be undeleted, why bother with Clone?

Reply:
I will make it private.

Reply:
Can't make it private because std::make_unique requires them to be accessible.

Reply:
The existence of Clone() still seems superfluous to me given that I can still make a copy of the object on the heap/stack with a public copy ctor. I suppose you can still return a unique_ptr using 'new' without making the copy ctor and ass op public. But I get that usage of 'new' is getting flagged. In that case, can we at least document in the header file that Clone() exists only to satisfy existing patterns?